### PR TITLE
nixos-option: fix build against nix 2.24

### PR DESCRIPTION
**Pull request by Mic92**
## Things done

I noticed today we still have nix 2.18 in our default nixos closure. Bumping nixos-option fixes that. This pull request just does the minimum to get nixos-option build.
I have already follow up commits to improve things further: https://github.com/NixOS/nixpkgs/pull/355748

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by fricklerhandwerk — APPROVED** [review]
LGTM at a cursory glance

File: pkgs/tools/nix/nixos-option/src/meson.build [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  'libnix-copy-paste.cc',
+]
+
+cc = meson.get_compiler('cpp')
+
+autotool_config = configuration_data()
+# nix defines this with autotools
+if cc.has_member('struct dirent', 'd_type', prefix: '#include <dirent.h>')
+  autotool_config.set('HAVE_STRUCT_DIRENT_D_TYPE', 1)
+endif
```
Comment (Ericson2314) [inline pkgs/tools/nix/nixos-option/src/meson.build:12]: I think if you `-include` the right `config.h`, it is fine. That should also be made less hacky upstream.

Comment (Ericson2314) [reply pkgs/tools/nix/nixos-option/src/meson.build:12]: See https://github.com/NixOS/hydra/blob/f974891c76e295240017dd7f04d50ecb4b70284e/configure.ac#L56

**Review by Ericson2314 — COMMENTED** [review]
Otherwise looks good!

File: pkgs/tools/nix/nixos-option/src/autotools-config.h.in
Comment (Ericson2314) [inline pkgs/tools/nix/nixos-option/src/autotools-config.h.in:1]: Can remove this file

**Review by Ericson2314 — APPROVED** [review]

**Comment by foolnotion** [comment]
This breaks the build on the current master. Last good commit is d13c9db8c31172de7dc424ca7d78fb08ed2dcc40

```bash
nix log /nix/store/0vj9pr9s98dkjv2imwdgp1vwfz5fnc8r-nixos-option.drv
warning: The interpretation of store paths arguments ending in `.drv` recently changed. If this command is now failing try again with '/nix/store/0vj9pr9s98dkjv2imwdgp1vwfz5fnc8r-nixos-option.drv^*'
Running phase: unpackPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking source archive /nix/store/gk2qs90xfhv0h19h8dca9hn9j3wkix6f-nixos-option
source root is nixos-option
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: configurePhase
@nix { "action": "setPhase", "phase": "configurePhase" }
mesonConfigurePhase flags: --prefix=/nix/store/z3bnffna084is69k0qn2jhimjshav2gm-nixos-option --libdir=/nix/store/z3bnffna084is69k0qn2jh>
The Meson build system
Version: 1.6.0
Source dir: /build/nixos-option
Build dir: /build/nixos-option/build
Build type: native build
Project name: nixos-option
Project version: 0.1.6
C++ compiler for the host machine: g++ (gcc 13.3.0 "g++ (GCC) 13.3.0")
C++ linker for the host machine: g++ ld.bfd 2.43.1
Host machine cpu family: x86_64
Host machine cpu: x86_64
Found pkg-config: YES (/nix/store/f1gamyykr0f8n4afy910ass2jx2g9zkv-pkg-config-wrapper-0.29.2/bin/pkg-config) 0.29.2
Run-time dependency nix-main found: YES 2.18.9
Run-time dependency nix-store found: YES 2.18.9
Run-time dependency nix-expr found: YES 2.18.9
Run-time dependency nix-cmd found: YES 2.18.9
Did not find CMake 'cmake'
Found CMake: NO
Run-time dependency nix-flake found: NO (tried pkgconfig)

meson.build:10:16: ERROR: Dependency "nix-flake" not found, tried pkgconfig

A full log can be found at /build/nixos-option/build/meson-logs/meson-log.txt
```

**Comment by lf-** [comment]
This effort would have been better spent on https://github.com/NixOS/nixpkgs/pull/313497 as nixos-option is a reimplementation of a bunch of the nix CLI front-end and is a giant pile of bad code. We should just get rid of it being a C++ API client and use the CLI and Nix expressions.

**Pull request merged**